### PR TITLE
DRAGEN: Coverage metrics update

### PR DESCRIPTION
**Pull request by Just-Roma**
New version of the DRAGEN coverage metrics module.  
It is aimed at the issue [1781](https://github.com/ewels/MultiQC/issues/1781), which is currently closed , but was not solved.

- The new code can detect all CSV files, which correspond to the following file structure:
  \<output prefix>.\<coverage region prefix>_coverage_metrics\<arbitrary suffix>.csv
- Part of the code automatically detects metrics and assigns some header's configs when possible.
- A special codeblock is introduced to give users a tool to easily refine header's configs manually.
- An extra module is added, which handles the overall_mean_cov.csv files. These contain the source bed/wgs files,
  which are included in the HTML report. No separate PR is created for that module, because it is used only by
  the DRAGEN Coverage module and does not play any role beyond that.
[test_coverage_data.zip](https://github.com/ewels/MultiQC/files/10944125/test_coverage_data.zip)

### CODE REVIEW

**Comment by your-highness** [comment]
Dear @ewels ,

It would be awesome if this module patch can _timely_ be integrated into main :smile:

Thanks

**Comment by ewels** [comment]
I will do my best.

@alexiswl - as you were the last person to work on this - if you could take a look at the code to help with the review process, that would be very much appreciated 🙏🏻

**Comment by your-highness** [comment]
Test Data can be found in ewels/MultiQC_TestData#248

**Comment by alexiswl** [comment]
> I will do my best.
> 
> @alexiswl - as you were the last person to work on this - if you could take a look at the code to help with the review process, that would be very much appreciated 🙏🏻

Yep happy to

**Comment by alexiswl** [comment]
## Checkout sources and build images

### Master Branch

```
mkdir -p src/master/
(
    cd src/master/;
    git clone -b master https://github.com/ewels/multiqc
    cd multiqc; 
    docker build -t localhost/multiqc:master .
)
```

### PR 1880 

```
mkdir -p src/pr1880/
(
    cd src/pr1880/;
    git clone -b Coverage-Metrics-Update https://github.com/just-roma/multiqc;
    cd multiqc; 
    docker build -t localhost/multiqc:pr1880 .
)
```


## Create test data

```
mkdir -p inputs/just_roma/test_data/
(
  cd inputs/just_roma/test_data/;
  wget "https://github.com/ewels/MultiQC/files/10944125/test_coverage_data.zip";
  unzip test_coverage_data.zip  
)
```

> UMCCR data as uploaded below (see review/test_data)

## Run images against just roma test data

### Master
```
mkdir -p outputs/master/just_roma_data_test/
docker run --rm \
  --volume "$PWD/inputs/just_roma/:/inputs" \
  --volume "$PWD/outputs/master/just_roma_data_test/:/outputs" \
  localhost/multiqc:master \
    multiqc /inputs \
      --outdir /outputs \
      --filename "master-just-roma-data-test" \
      --comment "master-with-just-roma-data-test"
```

### PR1880
```
mkdir -p outputs/pr1880/just_roma_data_test/
docker run --rm \
  --volume "$PWD/inputs/just_roma/:/inputs" \
  --volume "$PWD/outputs/pr1880/just_roma_data_test/:/outputs" \
  localhost/multiqc:pr1880 \
    multiqc /inputs \
      --outdir /outputs \
      --filename "pr1880-with-just-roma-data-test" \
      --comment "pr1880-with-just-roma-data-test"
```


## Run images against umccr data

### Master
```
mkdir -p outputs/master/umccr_data_test/
docker run --rm \
  --volume "$PWD/inputs/umccr_dragen/:/inputs" \
  --volume "$PWD/outputs/master/umccr_data_test/:/outputs" \
  localhost/multiqc:master \
    multiqc /inputs \
      --outdir /outputs \
      --filename "master-umccr-data-test" \
      --comment "master-with-umccr-data-test"
```

### PR1880
```
mkdir -p outputs/pr1880/umccr_data_test/
docker run --rm \
  --volume "$PWD/inputs/umccr_dragen/:/inputs" \
  --volume "$PWD/outputs/pr1880/umccr_data_test/:/outputs" \
  localhost/multiqc:pr1880 \
    multiqc /inputs \
      --outdir /outputs \
      --filename "pr1880-umccr-data-test" \
      --comment "pr1880-with-umccr-data-test"
```

## Review 

### Outputs

I've attached the four outputs below.  

* [master-just-roma-data-test.html.gz](https://github.com/ewels/MultiQC/files/10985587/master-just-roma-data-test.html.gz)
* [master-umccr-data-test.html.gz](https://github.com/ewels/MultiQC/files/10985592/master-umccr-data-test.html.gz)
* [pr1880-with-just-roma-data-test.html.gz](https://github.com/ewels/MultiQC/files/10985595/pr1880-with-just-roma-data-test.html.gz)
* [pr1880-umccr-data-test.html.gz](https://github.com/ewels/MultiQC/files/10985596/pr1880-umccr-data-test.html.gz)

### tl;dr 

Clearly a lot of work has gone into this, but I am skeptical of giving the green light 'as is'.  

### Test Data

The test data is very much appreciated and shows that the existing master branch doesn't handle target beds for germline samples (without a _tumor or _normal suffix). I am investigating that and have raised an issue [here](https://github.com/ewels/MultiQC/issues/1884) for this.  

I have also attached some 'umccr' test data that uses dragen version 4.0.3 

[umccr_dragen.zip](https://github.com/ewels/MultiQC/files/10985607/umccr_dragen.zip)

### Here are my main concerns for this PR:

 1. This PR does not handle somatic data (comparison below)
 2. There are too many columns showed by default.
 3. The default columns placed in the general statistics are not relevant to most users.
 4. Some odd commentary under names
 
### Somatic data issues

#### General Stats Page

The general statistics place the X cov WGS_normal in the tumor sample when run with PR1880 on somatic umccr data.

image

#### All other metrics are placed in their own header, rather than together 

i.e we have 'WGS_normal Metrics' instead of `WGS Metrics` containing both tumor and normal

**Master (with somatic umccr data)**

image

**PR1880 (with somatic umccr data) **

image

image

### Too many columns by default

#### Master

image

image

#### PR 1880

> It's all a bit overwhelming to look at, particularly when there are many samples

image
image

### General stats columns arent relevant

The general stats table can be a hot and crowded section for a multiqc report (particularly when many modules have been run on a sample).  

Currently the dragen module places just 'depth' in this section by default.  I don't think X cov WGS and Y cov WGS are the best default columns to place in this section

#### Master (with just roma data)

image


#### PR1880 (with just roma data)

image


### Some odd commentary under the names

**Master with just roma data**

image

_Coverage metrics over a region (where the region can be a target region, a QC coverage region, or the whole genome). Press the Help button for details.
_

**PR 1880 with just roma data**

_Coverage metrics over genome. Press the Help button for details._
_All_ samples are based on the Whole_genome_756_

image

**Comment by alexiswl** [comment]
> "This PR does not handle somatic data (comparison below)"

> Well, it does, just differently. Here is the output for 'umccr', but only for *_coverage_metrics.csv files.
> ...
> It is not a tumor sample, but just a sample.
> ...
> I have adjusted the code, so the output looks like from Master Branch now (but some 2Ds can not be created):
> ...
> I am not sure how important this 2D plotting for users though.

Splitting out the normal and tumor is quite important to us.  The tumor and normal are from different biopsies, completed in separate extractions, library preparations, adapters etc. While they are indeed of the same specimen, it cannot be said they are the same 'sample'.  

I think it is important that these can be split out as separate samples in all plots.  

> Two special additional header's configurations called "exclude" and "exclude_own" were introduced exactly for this
case.

Yes I was aware of this, however I think a user should be able to run the multiqc command with no tinkering and have it show a meaningful output without an overwhelming amount of columns.  The existing WGS metrics are probably a good place to start for this ( _Sample Name | M Aln reads | Mb Aln bases | Depth | Uniformity (>0.2×mean) | Mean/med autosomal coverage_).  

> "Some odd commentary under names"
> These are the file names extracted from the <sample name>.overall_mean_cov<tumor or normal>.csv files.
> The only reason why they are odd is because i thought them up.
> This is how it looks like for 'umccr' data:

Maybe this could be added to the help dropdown?

**Comment by ewels** [comment]
Agree - general pattern in MultiQC is to show the minimum number of columns that all users will want to see, then if needed have a config option to opt-in to additional columns. See for example, Picard.

**Comment by Just-Roma** [comment]
> Splitting out the normal and tumor is quite important to us. The tumor and normal are from different biopsies, completed in separate extractions, library preparations, adapters etc. While they are indeed of the same specimen, it cannot be said they are the same 'sample'.
> I think it is important that these can be split out as separate samples in all plots.

Done.

> The existing WGS metrics are probably a good place to start for this ( Sample Name | M Aln reads | Mb Aln bases | Depth | Uniformity (>0.2×mean) | Mean/med autosomal coverage).

New output for umccr (html attached below):



> Maybe this could be added to the help dropdown?

Included in the update.

[multiqc_report_umccr_PR1880.zip](https://github.com/ewels/MultiQC/files/11016660/multiqc_report_umccr_PR1880.zip)

**Comment by ewels** [comment]
Merging in the additional DRAGEN test data revealed a linting error, which is now making the CI fail on all other pull requests. It looks like you solved it in this PR as the tests pass, so merging this PR is now even higher priority 😅 

@alexiswl - let me know when you're happy and I'll have a look, then hopefully we can merge.

**Comment by alexiswl** [comment]
Sure will look at it tomorrow @ewels

**Comment by Just-Roma** [comment]
> Merging in the additional DRAGEN test data revealed a linting error, which is now making the CI fail on all other pull requests.

There were some strange fails in the log report a couple of times. Can not find them now after some of associated commits have been squashed. It had smth to do with fetching data, but it has failed only for a couple of OS tests: Windows 3.6/3.11 have failed, but others have passed. I assumed that there was some network issue and just restarted the checks. All have passed. 

PR1880 test data seems to have only 1 difference from other data sets: it has the README.md file. 
Maybe it's the root of the problem?

**Comment by alexiswl** [comment]
Looking much better! 

Looks like there's been a bit of change to the data structure under the hood too though that needs to be addressed. 

```
jq '.report_saved_raw_data | keys' < outputs/pr1880/umccr_data_test/pr1880-umccr-data-test_data/multiqc_data.json
[
  "dragen_cov_metrics",
  "dragen_frag_len",
  "dragen_map_metrics",
  "dragen_ploidy",
  "dragen_time_metrics",
  "dragen_trimmer_metrics",
  "dragen_vc_metrics",
  "multiqc_general_stats"
]
```

```
 jq '.report_saved_raw_data | .dragen_cov_metrics | keys' < outputs/pr1880/umccr_data_test/pr1880-umccr-data-test_data/multiqc_data.json
```

Yields

```
[
  "PRJ210678_normalqc-coverage-region-1",
  "PRJ210678_normalqc-coverage-region-2",
  "PRJ210678_normaltarget_bed",
  "PRJ210678_normalwgs",
  "PRJ210678qc-coverage-region-1",
  "PRJ210678qc-coverage-region-2",
  "PRJ210678target_bed",
  "PRJ210678wgs"
]
```

This should either be nested or should reflect to what exists in the master branch.

This is the outputs from the master branch

```
 jq '.report_saved_raw_data | keys' < outputs/master/umccr_data_test/master-umccr-data-test_data/multiqc_data.json
[
  "dragen_frag_len",
  "dragen_map_metrics",
  "dragen_ploidy",
  "dragen_qc_region_1_coverage_metrics",
  "dragen_qc_region_2_coverage_metrics",
  "dragen_target_bed_coverage_metrics",
  "dragen_time_metrics",
  "dragen_trimmer_metrics",
  "dragen_vc_metrics",
  "dragen_wgs_cov_metrics",
  "multiqc_general_stats"
]
```

```
jq '.report_saved_raw_data | .dragen_wgs_cov_metrics | keys' < outputs/master/umccr_data_test/master-umccr-data-test_data/multiqc_data.json
[
  "PRJ210678",
  "PRJ210678_normal"
]
```


@brainstorm @pdiakumis this PR would mean a large amount of changes would be required to the [dracarys](https://github.com/pdiakumis/dracarys) repository. 

Attached are the two multiqc data json files to compare side-by-side


[multiqc_data.prj210678.master.json.gz](https://github.com/ewels/MultiQC/files/11073435/multiqc_data.prj210678.master.json.gz)

[multiqc_data.prj210678.pr1880.json.gz](https://github.com/ewels/MultiQC/files/11073439/multiqc_data.prj210678.pr1880.json.gz)

**Comment by Just-Roma** [comment]
The code has been modified. This is how it looks now (data attached below):



The appearance of txt's internal data is slightly different though:
metrics are in lower case and '(second value)' string is used instead of 'pct' to produce a metric for the second value. 
If it is important for you, then please let me know.

[umccr_data.zip](https://github.com/ewels/MultiQC/files/11076516/umccr_data.zip)

**Comment by alexiswl** [comment]
@Just-Roma let me know when you've finished pushing to this branch and I'll reevaluate your upstream branch to the master branch

**Comment by Just-Roma** [comment]
@alexiswl this is the last update

**Comment by ewels** [comment]
Thanks for the updates @Just-Roma ! Things are a little slow with Easter, hope to get back to this soon..

**Comment by alexiswl** [comment]
Great work @Just-Roma! Have rerun the tests and all looks well from my side

**Comment by Just-Roma** [comment]
@your-highness, shall happen [soon](https://github.com/ewels/MultiQC/pull/1905)

**Comment by alexiswl** [comment]
Phil sorry I would like to have another glance over this since there's been changes to it since I last reviewed? 

@Just-Roma confirming there will be no more pushes to this branch? Also wondering why each push is a force-push?

**Comment by Just-Roma** [comment]
@alexiswl, yes this is the very last update. It most certainly does not affect the program's execution. Just modified the parameters (order and amount) of 2 logging utils. And excluded the "gen table" and "own table" substrings from metrics' IDs, since the gen table already gets some special substring, so there is no need to append those two. That's it.
Force-push? Just so there are fewer commits. Besides it seems that each commit is tested separately, after squashing they are all tested as a single group.

**Comment by alexiswl** [comment]
@ewels approved from my end

**Comment by ewels** [comment]
Thanks @alexiswl ! Will merge this ASAP.

File: multiqc/modules/dragen/utils.py [hunk 1]
```diff
@@ -141,3 +141,182 @@ def make_headers(parsed_metric_ids, metrics):
...
+    line1 = "\n  {} was built from the following samples:"
+    line2 = "\n    {} in {}"
+    for clean_sample_name in sample_names:
+        if len(sample_names[clean_sample_name]) > 1:
+            message += line1.format(clean_sample_name)
+            for file_handler in sample_names[clean_sample_name]:
+                message += line2.format(file_handler["fn"], file_handler["root"])
+    if message:
+        message = "\n\nDuplicate sample names were found. The last one overwrites previous data." + message
+        logger.warning(message)
```
Comment (ewels) [inline multiqc/modules/dragen/utils.py:158]: Please stick to the standard way of handling this for MultiQC, documented here: https://multiqc.info/docs/development/modules/#identical-sample-names

There are many reasons why duplicate sample names are found when running MultiQC, most are harmless. This warning log message is much too noisy.

**Review by ewels — APPROVED** [review]
Thanks for your work on this, much appreciated.

I think that at some point in the future, this module needs a complete rewrite from the ground up if I'm honest. But for now, fixing it and getting it up and running is certainly worthwhile and we can revisit the bigger picture issues in the future.

File: multiqc/modules/dragen/utils.py (hunk 1 above)
Comment (ewels) [reply multiqc/modules/dragen/utils.py:158]: Switched to `logger.debug` instead. Resolved
